renovate/src/Renovate/BinaryFormat/ELF/Internal.hs

```diff
@@ -0,0 +1,122 @@
+  then Nothing
+  else Just $ LoadSegmentInfo { pOffset =
+                                  let E.FileOffset off = E.phdrFileStart phdr
+                                  in off
+                              , pVAddr = E.phdrSegmentVirtAddr phdr
+                              , pMemSz = E.phdrMemSize phdr
+                              }
+
+-- | Find a spot in the program's virtual address space for the new PHDR segment
+--
+-- This is an implementation of the core of 'choosePHDRSegmentAddress', see the
+-- comment there for more detail.
+--
+-- This function assumes:
+--
+-- * There are no segments whose images overlap in the virtual address space
+--
+-- It guarantees that the address it returns is greater than or equal to the
+-- given offset (see commentary on 'doRewrite' in "Renovate.BinaryFormat.ELF").
+findSpaceForPHDRs ::
+  (Num (E.ElfWordType w), Ord (E.ElfWordType w), Integral (E.ElfWordType w)) =>
+  NEL.NonEmpty (LoadSegmentInfo w) {-^ Info about other LOAD segments -} ->
+  E.ElfWordType w {- ^ Offset of PHDR in file -} ->
+  E.ElfWordType w {- ^ Size of PHDR segment -} ->
+  Maybe (E.ElfWordType w)
+findSpaceForPHDRs segInfos phdrOffset phdrSize =
+  let sortedSegInfos@(firstSegment NEL.:| _) =
+        NEL.sortBy (O.comparing pVAddr) segInfos
+      lastSegment = NEL.head (NEL.reverse sortedSegInfos)
+      pgAlign = fromIntegral pageAlignment
+      -- The optimal address is either:
+      --
+      -- 1. before the first segment,
+      -- 2. after the last segment, or
+      -- 3. between two segments,
+      beforeFirst =
+        if pVAddr firstSegment > phdrSize + pgAlign
+        then Just $ alignValueDown (pVAddr firstSegment - phdrSize - 1) pgAlign
+        else Nothing
+      afterLast = Maybe.catMaybes
+        [ -- Immediately after
+          Just $ alignValue (pVAddr lastSegment + pMemSz lastSegment + 1) pgAlign
+        , -- Possibly much further after, if the provided PHDR segment offset is
+          -- higher than the range covered by the last segment.
+          if phdrOffset > pVAddr lastSegment + pMemSz lastSegment
+          then Just $ alignValue phdrOffset pgAlign
+          else Nothing
+        ]
+      -- We don't have to check every address between two segments, just the
+      -- maximal and minimal ones.
+      between =
+        let zipped = zip (NEL.toList sortedSegInfos)
+                         (drop 1 (NEL.toList sortedSegInfos))
+        in concat [ [ addr
+                    | (loSeg, hiSeg) <- zipped
+                    , let addr =
+                            alignValue (pVAddr loSeg + pMemSz loSeg + 1) pgAlign
+                    , phdrSize < pVAddr hiSeg - addr
+                    ]
+                  , [ addr
+                    | (loSeg, hiSeg) <- zipped
+                    , let addr =
+                            alignValueDown (pVAddr hiSeg - phdrSize - 1) pgAlign
+                    , addr > pVAddr loSeg + pMemSz loSeg
+                    ]
+                  ]
```

Comment:
This bit looks fairly subtle - can we factor out some bits with more descriptive names, or just give a textual description?

Reply:
I added some detailed commentary